feat(minifier): replace `const` with `let` for non-exported read-only variables

sapphi-red opened the pull request:
Replace `const` with `let` when that value does not have any assignments and not exposed.

sapphi-red commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8733?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8733** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8733?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8733?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8731** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8731?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8730** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8730?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8733).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
esbuild does const replacements, I'd like to do that at some point too.

So this should go to late pass instead of pre pass?

sapphi-red commented:
> esbuild does const replacements, I'd like to do that at some point too.
> 
> So this should go to late pass instead of pre pass?

I first implemented in the late pass and it wasn't idempotent because collapse_variable_declarations can merge `const a = 1; let b = 2;` into `let a = 1, b = 2`.
The information that it was a `const` can be obtained from `SymbolFlags::ConstVariable`. Would that work for const replacements?

Boshen commented:
> > esbuild does const replacements, I'd like to do that at some point too.
> > So this should go to late pass instead of pre pass?
> 
> I first implemented in the late pass and it wasn't idempotent because collapse_variable_declarations can merge `const a = 1; let b = 2;` into `let a = 1, b = 2`. The information that it was a `const` can be obtained from `SymbolFlags::ConstVariable`. Would that work for const replacements?

Yes, we can merge this PR regardless.

Boshen commented:
### Merge activity

* **Jan 26, 7:35 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
feat(minifier): replace `const` with `let` for non-exported read-only variables (#8733)

Replace `const` with `let` when that value does not have any assignments and not exposed.
